ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationQueue.java

```diff
@@ -0,0 +1,204 @@
+        && !queuedRegenerations.isEmpty()) {
+      if (activeRegenerations.compareAndSet(currentActiveCount, currentActiveCount + 1)) {
+        processNext();
+      }
+      currentActiveCount = activeRegenerations.get();
+    }
+  }
+
+  private void processNext() {
+    final RegenerationTask task = queuedRegenerations.poll();
+    if (task == null) {
+      activeRegenerations.decrementAndGet();
+      return;
+    }
+    task.regenerate()
+        .whenComplete(
+            (result, error) -> {
+              final SafeFuture<SignedBlockAndState> future =
+                  inProgressGeneration.remove(task.getBlockRoot());
+              if (error != null) {
+                future.completeExceptionally(error);
+              } else {
+                future.complete(result);
+              }
+            })
+        .alwaysRun(
+            () -> {
+              activeRegenerations.decrementAndGet();
+              tryProcessNext();
+            })
+        .reportExceptions();
+  }
+
+  public static class RegenerationTask {
+    private static final Logger LOG = LogManager.getLogger();
+    private final HashTree tree;
+    private final SignedBlockAndState baseBlockAndState;
+    private final BlockProvider blockProvider;
+    private final Bytes32 blockRoot;
+    private final Consumer<SignedBlockAndState> cacheHandler;
+
+    public RegenerationTask(
+        final Bytes32 blockRoot,
+        final HashTree tree,
+        final SignedBlockAndState baseBlockAndState,
+        final BlockProvider blockProvider,
+        final Consumer<SignedBlockAndState> cacheHandler) {
+      this.tree = tree;
+      this.baseBlockAndState = baseBlockAndState;
+      this.blockProvider = blockProvider;
+      this.blockRoot = blockRoot;
+      this.cacheHandler = cacheHandler;
+    }
+
+    public Bytes32 getBlockRoot() {
+      return blockRoot;
+    }
+
+    public HashTree getTree() {
+      return tree;
+    }
+
+    public RegenerationTask rebase(final SignedBlockAndState newBaseBlockAndState) {
+      final Bytes32 newBaseRoot = newBaseBlockAndState.getRoot();
+      if (!tree.contains(newBaseRoot)) {
+        LOG.warn(
+            "Attempting to rebase a task for {} onto a starting state that is not a required ancestor ({} at slot {})",
+            blockRoot,
+            newBaseRoot,
+            newBaseBlockAndState.getSlot());
+        return this;
+      }
+      final HashTree treeFromAncestor =
+          tree.withRoot(newBaseRoot).block(newBaseBlockAndState.getBlock()).build();
+      return new RegenerationTask(
+          blockRoot, treeFromAncestor, newBaseBlockAndState, blockProvider, cacheHandler);
+    }
+
+    public SafeFuture<SignedBlockAndState> regenerate() {
+      final StateGenerator stateGenerator =
+          StateGenerator.create(tree, baseBlockAndState, blockProvider);
+      return stateGenerator.regenerateStateForBlock(blockRoot).thenPeek(cacheHandler);
```

Comment:
Something to potentially experiment with in a follow-up PR:  we could use [`AsyncChainStateGenerator.generateStates(targetRoot, stateHandler)`](https://github.com/PegaSysEng/teku/blob/846737706485cd0a79ec1e6b1a3d56b7386e0e93/ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/AsyncChainStateGenerator.java#L85) to expose an API that gives us access to all intermediate states as well:  `StateGenerator.regenerateStatesForBlock(targetRoot, stateHandler)`.  We could then dedupe requests across the entire chain being processed.

Reply:
Interesting idea, the one concern I'd have is making sure we didn't consume too much memory in the process.